### PR TITLE
Rework build script

### PR DESCRIPTION
**Pull request by sleeepyjack**
This PR updates the `ci/build.sh` script to be usable both interactively and in CI.

### CODE REVIEW

**Comment by sleeepyjack** [comment]
Signature:
```shell
Usage: ci/build.sh [OPTIONS]

Options:
  -t/--tests: Build tests
  -e/--examples: Build examples
  -b/--benchmarks: Build benchmarks
  -c/--clean: Clean (re-)build
  --prefix: Build directory prefix (Defaults to <repo_root>/build)
  -i/--infix: Build directory infix (Defaults to local)
  -d/--debug: Debug build
  -p/--parallel: Build parallelism (Defaults to $PARALLEL_LEVEL if set, otherwise the system's number of CPU cores)
  --cuda: CUDA compiler (Defaults to $CUDACXX if set, otherwise nvcc)
  --cxx: Host compiler (Defaults to $CXX if set, otherwise g++)
  --arch: Target CUDA arches, e.g. "60-real;70;80-virtual" (Defaults to the system's native GPU archs)
  --std: CUDA/C++ standard (Defaults to 17)
  -v/-verbose/--verbose: Enable shell echo for debugging
  -h/-help/--help: Show this usage message

Examples:
  $ PARALLEL_LEVEL=8 CXX=g++-9 ci/build.sh
  $ ci/build.sh --cxx g++-9 -t --examples -p 8 -i my_build --debug
  $ ci/build.sh --cxx g++-8 --prefix /usr/local --arch 80-real -v --cuda /usr/local/bin/nvcc
```

**Comment by sleeepyjack** [comment]
Synopsis:
```
cuCollections build script
Usage: ci/build.sh [OPTIONS]
Options:
  -t/--tests: Build tests
  -e/--examples: Build examples
  -b/--benchmarks: Build benchmarks
  -c/--clean: Clean (re-)build
  --prefix: Build directory prefix (Defaults to <repo_root>/build)
  -i/--infix: Build directory infix (Defaults to local)
  -d/--debug: Debug build
  -p/--parallel: Build parallelism (Defaults to $PARALLEL_LEVEL if set, otherwise the system's number of CPU cores)
  --cuda: CUDA compiler (Defaults to $CUDACXX if set, otherwise nvcc)
  --cxx: Host compiler (Defaults to $CXX if set, otherwise g++)
  --arch: Target CUDA arches, e.g. "60-real;70;80-virtual" (Defaults to the system's native GPU archs)
  --std: CUDA/C++ standard (Defaults to 17)
  -v/-verbose/--verbose: Enable shell echo for debugging
  -h/-help/--help: Show this usage message

Examples:
  Basic Build:
    $ ci/build.sh
    Runs a basic build with default settings, i.e., builds tests, examples, and benchmarks.
    Build files will be written to buil/local and symlinked to build/latest.

  Debug Build with Tests and Examples:
    $ CXX=g++-9 ci/build.sh -t -e -d
    $ ci/build.sh --cxx g++-9 -t -e -d
    Sets the host compiler to g++-9, builds tests and examples, and enables debug mode.
    Build files will be written to build/local and symlinked to build/latest.

  Custom Build Directory with Benchmarks:
    $ BUILD_BENCHMARKS=ON ci/build.sh --prefix /custom/build --infix my_build
    $ ci/build.sh --prefix /custom/build --infix my_build -b
    Build files will be written to build/my_build and symlinked to build/latest.

  Custom Build Infix Directory:
    $ ci/build.sh -i my_build
    Builds with benchmarks in the /custom/build/my_build directory.

  Parallel Build with Specific CUDA Architecture and CUDA Compiler:
    $ PARALLEL_LEVEL=8 ci/build.sh --cuda /my_cuda_compiler/nvcc --arch 70;80
    $ ci/build.sh -p 8 --cuda /my_cuda_compiler/nvcc --arch 70;80
    Specifies parallel build level of 8 and CUDA architecture 70 and 80 with the specified CUDA compiler.

  Verbose Mode for Debugging:
    $ ci/build.sh -v --std 17
    Enables verbose mode for detailed build process output and build with C++17 standard.
```

**Comment by sleeepyjack** [comment]
Synopsis:
```
cuCollections build script
Usage: ci/build.sh [OPTIONS]
Options:
  -t/--tests: Build tests
  -e/--examples: Build examples
  -b/--benchmarks: Build benchmarks
  -c/--clean: Clean (re-)build
  --prefix: Build directory prefix (Defaults to <repo_root>/build)
  -i/--infix: Build directory infix (Defaults to local)
  -d/--debug: Debug build
  -p/--parallel: Build parallelism (Defaults to $PARALLEL_LEVEL if set, otherwise the system's number of CPU cores)
  --cuda: CUDA compiler (Defaults to $CUDACXX if set, otherwise nvcc)
  --cxx: Host compiler (Defaults to $CXX if set, otherwise g++)
  --arch: Target CUDA arches, e.g. "60-real;70;80-virtual" (Defaults to the system's native GPU archs)
  --std: CUDA/C++ standard (Defaults to 17)
  -v/-verbose/--verbose: Enable shell echo for debugging
  -h/-help/--help: Show this usage message

Examples:
  Basic Build:
    $ ci/build.sh
    Runs a basic build with default settings, i.e., builds tests, examples, and benchmarks.
    Build files will be written to <repo_root>/build/local and symlinked to <repo_root>/build/latest.

  Debug Build with Tests and Examples:
    $ CXX=g++-9 ci/build.sh -t -e -d
    $ ci/build.sh --cxx g++-9 -t -e -d
    Sets the host compiler to g++-9, builds tests and examples, and enables debug mode.
    Build files will be written to <repo_root>/build/local and symlinked to <repo_root>/build/latest.

  Custom Build Directory with Benchmarks:
    $ BUILD_BENCHMARKS=ON ci/build.sh --prefix /custom/build --infix my_build
    $ ci/build.sh --prefix /custom/build --infix my_build -b
    Build files will be written to /custom/build/my_build and symlinked to /custom/build/latest.

  Custom Build Infix Directory:
    $ ci/build.sh -i my_build
    Builds with benchmarks in the <repo_root>/build/my_build directory and symlinked to <repo_root>/build/latest.

  Parallel Build with Specific CUDA Architecture and CUDA Compiler:
    $ PARALLEL_LEVEL=8 ci/build.sh --cuda /my_cuda_compiler/nvcc --arch 70;80
    $ ci/build.sh -p 8 --cuda /my_cuda_compiler/nvcc --arch 70;80
    Specifies parallel build level of 8 and CUDA architecture 70 and 80 with the specified CUDA compiler.
    Build files will be written to <repo_root>/build/local and symlinked to <repo_root>/build/latest.

  Verbose Mode for Debugging:
    $ ci/build.sh -v --std 17
    Enables verbose mode for detailed build process output and build with C++17 standard.
    Build files will be written to <repo_root>/build/local and symlinked to <repo_root>/build/latest.
```

File: README.md [hunk 1]
```diff
@@ -59,8 +59,7 @@ This will take care of downloading `cuCollections` from GitHub and making the he
 
 `cuCollections` depends on the following libraries:
 
-- [libcu++](https://github.com/NVIDIA/libcudacxx)
-- [CUB](https://github.com/thrust/cub)
+- [CUDA C++ Core Libraries (CCCL)](https://github.com/NVIDIA/cccl)
```
Comment (PointKernel) [inline README.md:62]: we are not using `cccl` yet, right?

File: README.md
Comment (PointKernel) [inline README.md:1]: could you add a sub-section under https://github.com/NVIDIA/cuCollections#building-cucollections explaining how to build with the script?

File: README.md (hunk 1 above)
Comment (sleeepyjack) [reply README.md:62]: The answer is complicated I guess. Technically speaking, all of the cccl facilities are now unified under this repo and the separate repos are deprecated. I know rapids-cmake is still in the process of migrating to a uniform version of all cccl libs, but that's a temporary detail. From a user's perspective this is more clear.

File: README.md
Comment (sleeepyjack) [reply README.md:1]: I can paste the example from the help message

File: README.md (hunk 1 above)
Comment (PointKernel) [reply README.md:62]: okay. that's fair

File: README.md
Comment (sleeepyjack) [reply README.md:1]: done

File: README.md [hunk 2]
```diff
@@ -80,9 +79,65 @@ make
...
+    $ PARALLEL_LEVEL=8 ci/build.sh --cuda /my_cuda_compiler/nvcc --arch 70;80
+    $ ci/build.sh -p 8 --cuda /my_cuda_compiler/nvcc --arch 70;80
+    Specifies parallel build level of 8 and CUDA architecture 70 and 80 with the specified CUDA compiler.
+    Build files will be written to <repo_root>/build/local and symlinked to <repo_root>/build/latest.
+
+  Verbose Mode for Debugging:
+    $ ci/build.sh -v --std 17
+    Enables verbose mode for detailed build process output and build with C++17 standard.
+    Build files will be written to <repo_root>/build/local and symlinked to <repo_root>/build/latest.
+```
```
Comment (PointKernel) [inline README.md]: those are the details not necessarily needed at the top-level README.
```suggestion

```

**Review by PointKernel — APPROVED** [review]
LGTM

Just one non-blocking suggestion.

File: README.md (hunk 2 above)
Comment (sleeepyjack) [reply README.md]: Agreed. I updated the readme like so: [143df1d](https://github.com/NVIDIA/cuCollections/pull/395/commits/143df1d93867dddc09f96306afcaeb62d613d9ad)

Comment (PointKernel) [reply README.md]: Looks great!

**Pull request merged**